dissect/target/plugins/apps/webservers/caddy.py

```diff
@@ -0,0 +1,128 @@
+        if (config_file := self.target.fs.path("/etc/caddy/Caddyfile")).exists():
+            found_roots = []
+            for line in config_file.open("rt"):
+                line = line.strip()
+
+                if not line or line.startswith("#"):
+                    continue
+
+                if "root " in line:
+                    found_roots.append(line.split("root ")[1].strip())
+
+                if "log " in line or "output file " in line:
+                    if line.strip() == "log {":
+                        continue
+
+                    match = LOG_FILE_REGEX.match(line)
+                    if not match:
+                        self.target.log.warning("Could not determine log path in line %s", line)
+                        continue
+
+                    match = match.groupdict()
+                    p = match["log_file"].replace(" {", "").split(" ")[-1]
+
+                    if not p.startswith("/"):
+                        print(f"{p} is a relative path")
+                        for root in found_roots:
+                            root_parent = self.target.fs.path(f"{root}").parent
+                            abs_log_path = self.target.fs.path(f"{str(root_parent)}/{p}").parent
+                            print(abs_log_path)
+                            if abs_log_path.exists() and abs_log_path not in log_paths:
+                                log_paths.append(abs_log_path)
+                    else:
+                        print(f"{p} is an absolute path")
+                        if (path := self.target.fs.path(p).parent).exists():
+                            if path not in log_paths:
+                                log_paths.append(path)
+
+        return log_paths
+
+    def check_compatible(self) -> bool:
+        return len(self.log_paths) > 0
+
+    @plugin.export(record=WebserverRecord)
+    def access(self) -> Iterator[WebserverRecord]:
+        """Parses Caddy V1 CRF and Caddy V2 JSON logs.
+
+        Resources:
+            - https://caddyserver.com/docs/caddyfile/directives/log#format-modules
+        """
+        tzinfo = self.target.datetime.tzinfo
+
+        for path in self.log_paths:
+            for line in open_decompress(path, "rt"):
+                line = line.strip()
+                if not line:
+                    continue
+
+                # Parse a JSON log line
+                if line.startswith('{"'):
+                    try:
+                        log = json.loads(line)
+                    except json.decoder.JSONDecodeError:
+                        self.target.log.warning("Could not decode Caddy JSON log line in file %s", path)
+                        continue
+
+                    yield WebserverRecord(
+                        ts=from_unix(log["ts"]).replace(tzinfo=tzinfo),
+                        remote_ip=log["request"]["remote_ip"],
+                        url=f"{log['request']['method']} {log['request']['uri']} {log['request']['proto']}",
```

Comment:
I assume `log['request']['method']` is strings like `GET`, `POST` etc, and `log['request']['proto']` is `HTTP/2`?

I'd suggest putting just the `log['request']['uri']` in the `url` field of the record, and either create new fields for the `proto` and `method` if you want to include them.

Reply:
This is to ensure uniformity across all other webserver log parsers. The format for nginx, apache and caddy CLF is always `method uri proto`.

Reply:
I changed it to separate `method`, `uri` and `protocol` fields.